bindings/go/generator/jsonschemagen/schema.go

```diff
@@ -0,0 +1,367 @@
+		sch := &Schema{
+			Schema:               "https://json-schema.org/draft/2020-12/schema",
+			Comment:              "generated by the ocm schema generation tool",
+			ID:                   g.schemaID(ti),
+			Title:                ti.Key.TypeName,
+			Description:          desc,
+			Type:                 "object",
+			Properties:           g.buildStructProperties(ti.Struct, ti),
+			Required:             g.buildStructRequired(ti.Struct),
+			AdditionalProperties: &SchemaOrBool{Bool: Ptr(false)},
+		}
+
+		if deprecated {
+			sch.Deprecated = Ptr(true)
+		}
+		return sch
+	}
+
+	return g.buildAliasSchema(ti)
+}
+
+func (g *Generator) buildAliasSchema(ti *universe.TypeInfo) *Schema {
+	desc, deprecated := extractStructDoc(ti.TypeSpec, ti.GenDecl)
+
+	switch t := ti.Expr.(type) {
+	case *ast.Ident: // type Foo string
+		if prim := primitiveSchema(t.Name); prim != nil {
+			s := &Schema{
+				Schema:      "https://json-schema.org/draft/2020-12/schema",
+				Comment:     "generated by the ocm schema generation tool",
+				ID:          g.schemaID(ti),
+				Title:       ti.Key.TypeName,
+				Description: desc,
+				Type:        prim.Type,
+			}
+			if deprecated {
+				s.Deprecated = Ptr(true)
+			}
+			// (optional future) enum extraction
+			return s
+		}
+
+	case *ast.ArrayType: // type Foo []Bar
+		return &Schema{
+			Schema:  "https://json-schema.org/draft/2020-12/schema",
+			Comment: "generated by the ocm schema generation tool",
+			ID:      g.schemaID(ti),
+			Title:   ti.Key.TypeName,
+			Type:    "array",
+			Items:   g.schemaForExpr(t.Elt, ti),
+		}
+
+	case *ast.MapType: // type Foo map[string]Bar
+		return &Schema{
+			Schema:  "https://json-schema.org/draft/2020-12/schema",
+			Comment: "generated by the ocm schema generation tool",
+			ID:      g.schemaID(ti),
+			Title:   ti.Key.TypeName,
+			Type:    "object",
+			AdditionalProperties: &SchemaOrBool{
+				Schema: g.schemaForExpr(t.Value, ti),
+			},
+		}
+	}
+
+	// fallback
+	return anyObjectSchema()
+}
+
+func (g *Generator) buildStructProperties(st *ast.StructType, ti *universe.TypeInfo) map[string]*Schema {
+	props := map[string]*Schema{}
+
+	for _, field := range st.Fields.List {
+		if len(field.Names) == 0 {
+			continue
+		}
+
+		jsonName := jsonTagName(field, field.Names[0].Name)
+		if jsonName == "-" {
+			continue
+		}
+
+		sch := g.schemaForExpr(field.Type, ti)
+
+		desc, deprecated := extractFieldDoc(field)
+		if desc != "" {
+			sch.Description = desc
+		}
+		if deprecated {
+			sch.Deprecated = Ptr(true)
+		}
+
+		props[jsonName] = sch
+	}
+
+	return props
+}
+
+func (g *Generator) buildStructRequired(st *ast.StructType) []string {
+	var req []string
+	for _, f := range st.Fields.List {
+		if len(f.Names) == 0 {
+			continue
+		}
+		if !jsonTagHasOmitEmpty(f) {
+			req = append(req, jsonTagName(f, f.Names[0].Name))
+		}
+	}
+	return req
+}
+
+func (g *Generator) schemaForExpr(expr ast.Expr, ctx *universe.TypeInfo) *Schema {
+	switch t := expr.(type) {
+	case *ast.Ident:
```

Comment:
If we include`*ast.InterfaceType`, we could get the actual implementing type of the thing and display that instead of a plain `Object`. WDYT?

Reply:
Okay, fair point.